tests/translate/convert/test_rc2po.py failures with Python 3.12

When running the tests with Python 3.12, I'm getting the following 4 test failures:

```
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert_encoding_utf16 - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert_encoding_utf8 - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_menuex - assert 0 == 5
```

The respective test pass with Python 3.11.6.

Reproducer and output:

```pytb
$ python3.12 -m venv .venv
$ . .venv/bin/activate
$ pip install -q . pytest iniparse vobject mistletoe ruamel.yaml pyparsing phply
$ python -m pytest tests/translate/convert/
========================================================= test session starts =========================================================
platform linux -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0
rootdir: /tmp/translate
configfile: pytest.ini
collected 1067 items / 1 skipped                                                                                                      

tests/translate/convert/test_accesskey.py ............                                                                          [  1%]
tests/translate/convert/test_android2po.py ......                                                                               [  1%]
tests/translate/convert/test_convert.py .                                                                                       [  1%]
tests/translate/convert/test_csv2po.py ...................                                                                      [  3%]
tests/translate/convert/test_dtd2po.py ........................x............................x.....                              [  9%]
tests/translate/convert/test_flatxml2po.py .........                                                                            [  9%]
tests/translate/convert/test_html2po.py ....................................................................................... [ 18%]
.......                                                                                                                         [ 18%]
tests/translate/convert/test_ical2po.py .......................                                                                 [ 20%]
tests/translate/convert/test_idml2po.py ..                                                                                      [ 21%]
tests/translate/convert/test_ini2po.py .................                                                                        [ 22%]
tests/translate/convert/test_json2po.py .......                                                                                 [ 23%]
tests/translate/convert/test_md2po.py .....                                                                                     [ 23%]
tests/translate/convert/test_moz2po.py .                                                                                        [ 23%]
tests/translate/convert/test_mozfunny2prop.py ..                                                                                [ 24%]
tests/translate/convert/test_mozlang2po.py .................                                                                    [ 25%]
tests/translate/convert/test_ods2xliff.py ..                                                                                    [ 25%]
tests/translate/convert/test_oo2po.py ......................                                                                    [ 27%]
tests/translate/convert/test_oo2xliff.py .....................                                                                  [ 29%]
tests/translate/convert/test_php2po.py ...................................                                                      [ 33%]
tests/translate/convert/test_po2csv.py .....................                                                                    [ 35%]
tests/translate/convert/test_po2dtd.py .........................................................                                [ 40%]
tests/translate/convert/test_po2flatxml.py .........                                                                            [ 41%]
tests/translate/convert/test_po2html.py ..................................                                                      [ 44%]
tests/translate/convert/test_po2ical.py .........................                                                               [ 46%]
tests/translate/convert/test_po2idml.py ..                                                                                      [ 47%]
tests/translate/convert/test_po2ini.py ...............................                                                          [ 49%]
tests/translate/convert/test_po2json.py .......                                                                                 [ 50%]
tests/translate/convert/test_po2md.py ....                                                                                      [ 50%]
tests/translate/convert/test_po2moz.py .                                                                                        [ 51%]
tests/translate/convert/test_po2mozlang.py .....................                                                                [ 53%]
tests/translate/convert/test_po2oo.py .....................                                                                     [ 55%]
tests/translate/convert/test_po2php.py ..................x..................x.                                                  [ 58%]
tests/translate/convert/test_po2prop.py ...................................................                                     [ 63%]
tests/translate/convert/test_po2rc.py ............                                                                              [ 64%]
tests/translate/convert/test_po2resx.py .....................................                                                   [ 68%]
tests/translate/convert/test_po2tiki.py .......                                                                                 [ 68%]
tests/translate/convert/test_po2tmx.py ...........................                                                              [ 71%]
tests/translate/convert/test_po2ts.py .................                                                                         [ 72%]
tests/translate/convert/test_po2txt.py .................                                                                        [ 74%]
tests/translate/convert/test_po2web2py.py ....                                                                                  [ 74%]
tests/translate/convert/test_po2xliff.py ................                                                                       [ 76%]
tests/translate/convert/test_po2yaml.py .....................                                                                   [ 78%]
tests/translate/convert/test_pot2po.py .....x...xx............................x...xx........................                    [ 84%]
tests/translate/convert/test_prop2mozfunny.py ...                                                                               [ 85%]
tests/translate/convert/test_prop2po.py .............................................                                           [ 89%]
tests/translate/convert/test_rc2po.py .FF.FF                                                                                    [ 89%]
tests/translate/convert/test_resx2po.py ..............                                                                          [ 91%]
tests/translate/convert/test_tiki2po.py .......                                                                                 [ 91%]
tests/translate/convert/test_ts2po.py .................                                                                         [ 93%]
tests/translate/convert/test_txt2po.py .........................                                                                [ 95%]
tests/translate/convert/test_web2py2po.py ...                                                                                   [ 95%]
tests/translate/convert/test_xliff2odf.py ..                                                                                    [ 96%]
tests/translate/convert/test_xliff2po.py ............................                                                           [ 98%]
tests/translate/convert/test_yaml2po.py ....x.....x..                                                                           [100%]

============================================================== FAILURES ===============================================================
____________________________________________________ TestRC2POCommand.test_convert ____________________________________________________

self = <tests.translate.convert.test_rc2po.TestRC2POCommand object at 0x7fd7d80606b0>

    def test_convert(self):
        """Tests the conversion to a po file"""
        self.create_testfile("simple.rc", RC_SOURCE)
        self.run_command(i="simple.rc", o="simple.po")
        po_result = pofile(self.open_testfile("simple.po"))
>       assert len(po_result.units) == 23
E       assert 0 == 23
E        +  where 0 = len([])
E        +    where [] = <translate.storage.pypo.pofile object at 0x7fd7d7cbbe60>.units

tests/translate/convert/test_rc2po.py:92: AssertionError
____________________________________________ TestRC2POCommand.test_convert_encoding_utf16 _____________________________________________

self = <tests.translate.convert.test_rc2po.TestRC2POCommand object at 0x7fd7d8060290>

    def test_convert_encoding_utf16(self):
        self.create_testfile("simple.rc", RC_SOURCE.encode("utf-16"))
        self.run_command(i="simple.rc", o="simple.po", charset="utf-16")
        po_result = pofile(self.open_testfile("simple.po"))
>       assert len(po_result.units) == 23
E       assert 0 == 23
E        +  where 0 = len([])
E        +    where [] = <translate.storage.pypo.pofile object at 0x7fd7d7a5dc70>.units

tests/translate/convert/test_rc2po.py:108: AssertionError
_____________________________________________ TestRC2POCommand.test_convert_encoding_utf8 _____________________________________________

self = <tests.translate.convert.test_rc2po.TestRC2POCommand object at 0x7fd7d80622a0>

    def test_convert_encoding_utf8(self):
        self.create_testfile("simple.rc", RC_SOURCE.encode("utf-8"))
        self.run_command(i="simple.rc", o="simple.po", charset="utf-8")
        po_result = pofile(self.open_testfile("simple.po"))
>       assert len(po_result.units) == 23
E       assert 0 == 23
E        +  where 0 = len([])
E        +    where [] = <translate.storage.pypo.pofile object at 0x7fd7d798c290>.units

tests/translate/convert/test_rc2po.py:123: AssertionError
____________________________________________________ TestRC2POCommand.test_menuex _____________________________________________________

self = <tests.translate.convert.test_rc2po.TestRC2POCommand object at 0x7fd7d8061970>

        def test_menuex(self):
            source = """
    LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_US
    
    IDM_STARTMENU MENUEX
    BEGIN
        POPUP ""
        BEGIN
            MENUITEM "", -1, MFT_SEPARATOR
            POPUP "&Programs", IDM_PROGRAMS
            BEGIN
                MENUITEM "(Empty)", -1, MFT_STRING, MFS_GRAYED
            END
            MENUITEM "Sh&ut Down...", IDM_SHUTDOWN, MFT_STRING, MFS_ENABLED
        END
    END
    """
            self.create_testfile("simple.rc", source.encode("utf-8"))
            self.run_command(i="simple.rc", o="simple.po", charset="utf-8")
            po_result = pofile(self.open_testfile("simple.po"))
>           assert len(po_result.units) == 5
E           assert 0 == 5
E            +  where 0 = len([])
E            +    where [] = <translate.storage.pypo.pofile object at 0x7fd7d7967380>.units

tests/translate/convert/test_rc2po.py:145: AssertionError
========================================================== warnings summary ===========================================================
.venv/lib/python3.12/site-packages/dateutil/tz/tz.py:37
  Warning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).

tests/translate/convert/test_po2dtd.py::TestPO2DTD::test_accesskey_missing
tests/translate/convert/test_po2dtd.py::TestPO2DTD::test_accesskey_and_amp_case_no_accesskey
tests/translate/convert/test_po2dtd.py::TestPO2DTD::test_accesskey_and_amp_case_amp_no_accesskey
tests/translate/convert/test_po2dtd.py::TestPO2DTDCommand::test_accesskey_missing
tests/translate/convert/test_po2dtd.py::TestPO2DTDCommand::test_accesskey_and_amp_case_no_accesskey
tests/translate/convert/test_po2dtd.py::TestPO2DTDCommand::test_accesskey_and_amp_case_amp_no_accesskey
  Warning: Could not find accesskey for key.accesskey

tests/translate/convert/test_po2prop.py::TestPO2Prop::test_mozilla_accesskeys_missing_accesskey
tests/translate/convert/test_po2prop.py::TestPO2PropCommand::test_mozilla_accesskeys_missing_accesskey
  Warning: Could not find accesskey for prop.accesskey

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= short test summary info =======================================================
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert_encoding_utf16 - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_convert_encoding_utf8 - assert 0 == 23
FAILED tests/translate/convert/test_rc2po.py::TestRC2POCommand::test_menuex - assert 0 == 5
================================= 4 failed, 1051 passed, 1 skipped, 12 xfailed, 9 warnings in 12.89s ==================================
```

--- a/translate/misc/optrecurse.py
+++ b/translate/misc/optrecurse.py
@@ -684,10 +684,17 @@ class RecursiveOptionParser(optparse.OptionParser):
             tempoutput = False
         templatefile = self.opentemplatefile(options, fulltemplatepath)
         passthroughoptions = self.getpassthroughoptions(options)
-        if fileprocessor(inputfile, outputfile, templatefile, **passthroughoptions):
+        result = fileprocessor(
+            inputfile, outputfile, templatefile, **passthroughoptions
+        )
+        if fullinputpath is not None:
+            inputfile.close()
+        if result:
             if tempoutput:
                 self.warning("writing to temporary output...")
                 self.finalizetempoutputfile(options, outputfile, fulloutputpath)
+            if fulloutputpath and os.path.isfile(fulloutputpath):
+                outputfile.close()
             return True
         else:
             # remove the file if it is a file (could be stdout etc)

